### PR TITLE
ticket-3 much better async calls

### LINKED ISSUE #3: Improve asynchronous call handlers
Asynchronous call handlers handlers in IMA Agent, such as JSON RPC class, often occupy lot of nested code space because of compound call handlers. This is especially visible with multiple nested asynchronous call handlers.
Make asynchronous call handlers split into call initiators and handler functions defined separately from each other and allow to read code of IMA Agent in a more plain manner.

### PR DESCRIPTION
**Pull request by sergiy-skalelabs**
Refactored asynchronous calls like JSON RPC everywhere. Much less code indenting. Lot of big functions split into smaller ones.
No performance changes introduced.
No new tests needed.

### CODE REVIEW

File: src/bls.mjs [hunk 1]
```diff
@@ -911,7 +911,7 @@ async function gatherSigningStartImpl( optsSignOperation ) {
     optsSignOperation.details.debug( "{p}Waiting for BLS glue result...",
         optsSignOperation.strLogPrefix );
     optsSignOperation.errGathering = null;
-    optsSignOperation.promiseCompleteGathering = new Promise( ( resolve, reject ) => {
+    optsSignOperation.signaturesGatheringDone = new Promise( ( resolve, reject ) => {
         const iv = setInterval( function() {
```
Comment (DmytroNazarenko) [inline src/bls.mjs]: Let's move it outside of Promise

File: src/imaCore.mjs [hunk 2]
```diff
@@ -664,7 +664,7 @@ async function callbackAllMessagesSign( optsTransfer, err, jarrMessages, joGlueR
 
 async function handleAllMessagesSigning( optsTransfer ) {
```
Comment (DmytroNazarenko) [inline src/imaCore.mjs:665]: Convert to raw function

File: src/main.mjs [hunk 3]
```diff
@@ -421,8 +421,7 @@ async function main() {
     initMonitoringServer();
     initJsonRpcServer();
     const isSilentReDiscovery = imaState.isPrintSecurityValues
-        ? false
-        : imaState.joSChainDiscovery.isSilentReDiscovery;
+        ? false : imaState.joSChainDiscovery.isSilentReDiscovery;
     const fnOnPeriodicDiscoveryResultAvailable = function( isFinal ) {
         loop.spreadUpdatedSChainNetwork( isFinal );
     };
```
Comment (DmytroNazarenko) [inline src/main.mjs:427]: line 441: make it await

File: src/bls.mjs (hunk 1 above)
Comment (sergiy-skalelabs) [reply src/bls.mjs]: Done

File: src/imaCore.mjs (hunk 2 above)
Comment (sergiy-skalelabs) [reply src/imaCore.mjs:665]: Done

File: src/main.mjs (hunk 3 above)
Comment (sergiy-skalelabs) [reply src/main.mjs:427]: Done

File: src/discoveryTools.mjs
Comment (DmytroNazarenko) [inline src/discoveryTools.mjs:1]: Remove Promise from discoverSChainNetwork

File: src/oracle.mjs [hunk 4]
```diff
@@ -83,6 +80,97 @@ export function findPowNumber( strRequestPart, details, isVerbose ) {
...
+        }
+    }
+    details.error( "RPC call(oracle_checkResult) all attempts timed out" );
+    reject( new Error( "RPC call(oracle_checkResult) all attempts timed out" ) );
+    await joCall.disconnect();
+}
+
 export function oracleGetGasPrice( oracleOpts, details ) {
     details = details || log;
     const promiseComplete = new Promise( ( resolve, reject ) => {
```
Comment (DmytroNazarenko) [inline src/oracle.mjs]: Remove Promise

**Review by DmytroNazarenko — COMMENTED** [review]
* Move sleep function to separate file and import it where needed

File: src/discoveryTools.mjs
Comment (sergiy-skalelabs) [reply src/discoveryTools.mjs:1]: Done

File: src/oracle.mjs (hunk 4 above)
Comment (sergiy-skalelabs) [reply src/oracle.mjs]: Done

File: src/discoveryTools.mjs
Comment (sergiy-skalelabs) [reply src/discoveryTools.mjs:1]: Done

File: src/oracle.mjs (hunk 4 above)
Comment (sergiy-skalelabs) [reply src/oracle.mjs]: Done

File: src/clpTools.mjs [hunk 5]
```diff
@@ -1444,6 +1444,63 @@ export function commandLineTaskLoopSimple() {
...
+            log.critical( "{p}Discovery node {} is completely unknown and will be skipped",
+                strLogPrefix, i );
+            continue;
+        }
+        const strNodeURL = imaUtils.composeSChainNodeUrl( joNode );
+        const rpcCallOpts = null;
+        await rpcCall.create(
+            strNodeURL,
+            rpcCallOpts,
+            async function( joCall, err ) {
```
Comment (DmytroNazarenko) [inline src/clpTools.mjs:1473]: Can we retrieve this function out of `create`?

**Review by dmytrotkk — APPROVED** [review]

**Review by badrogger — APPROVED** [review]

File: src/main.mjs [hunk 6]
```diff
@@ -439,7 +438,7 @@ async function main() {
             ( isSilentReDiscovery
                 ? log.fmtWarning( "silent" ) : log.fmtSuccess( "exposed details" ) ) );
         if( ! imaState.bNoWaitSChainStarted ) {
-            discoveryTools.waitUntilSChainStarted().then( function() {
+            await discoveryTools.waitUntilSChainStarted().then( function() {
```
Comment (DmytroNazarenko) [inline src/main.mjs]: Then is unneeded

**Review by DmytroNazarenko — CHANGES_REQUESTED** [review]

File: src/main.mjs (hunk 6 above)
Comment (sergiy-skalelabs) [reply src/main.mjs]: Need some explanation. Did't got idea.

File: src/clpTools.mjs (hunk 5 above)
Comment (sergiy-skalelabs) [reply src/clpTools.mjs:1473]: We can, but I left it as is because it's 10 lines function.

File: src/main.mjs (hunk 6 above)
Comment (sergiy-skalelabs) [reply src/main.mjs]: Removed, modified

**Review by DmytroNazarenko — APPROVED** [review]

**Pull request merged**